File: reference-implementations/java/src/main/java/org/tomlschema/SchemaDefinition.java
package org.tomlschema;

import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

record SchemaDefinition(
        String name,
        SchemaType type,
        String reference,
        String description,
        String itemReference,
        List<String> items,
        boolean optional,
        List<Object> allowedValues,
        Pattern pattern,
        SchemaStringFormat format,
        Pattern keyPattern,
        Object min,
        Object max,
        Integer minLength,
        Integer maxLength,
        List<String> oneOf,
        List<String> anyOf,
        SchemaCondition condition,
        String thenReference,
        String elseReference,
        List<String> allOf,
        Map<String, List<String>> dependentRequired,
        List<List<String>> mutuallyExclusive,
        List<List<String>> exactlyOne,
        Boolean uniqueItems,
        boolean hasDefault,
        Object defaultValue,
        boolean deprecated,
        Map<String, SchemaDefinition> children
) {
    SchemaDefinition {
        allowedValues = List.copyOf(allowedValues);
        oneOf = List.copyOf(oneOf);
        anyOf = List.copyOf(anyOf);
        allOf = List.copyOf(allOf);
        dependentRequired = dependentRequired.entrySet().stream()
                .collect(java.util.stream.Collectors.toUnmodifiableMap(
                        Map.Entry::getKey, entry -> List.copyOf(entry.getValue())));
        mutuallyExclusive = mutuallyExclusive.stream().map(List::copyOf).toList();
        exactlyOne = exactlyOne.stream().map(List::copyOf).toList();
        items = List.copyOf(items);
        children = Map.copyOf(children);
    }

}

File: reference-implementations/dotnet/src/SchemaDefinition.cs
namespace TomlSchema;

/// <summary>
/// Represents a schema definition node (element or type).
/// </summary>
public class SchemaDefinition
{
    public SchemaType Type { get; set; }
    public string? ItemType { get; set; }
    public string? Description { get; set; }
    public bool Optional { get; set; }
    public object? DefaultValue { get; set; }
    public bool Deprecated { get; set; }
    public string? Pattern { get; set; }
    public string? KeyPattern { get; set; }
    public long? Min { get; set; }
    public long? Max { get; set; }
    public long? MinLength { get; set; }
    public long? MaxLength { get; set; }
    public bool? UniqueItems { get; set; }
    public List<object>? AllowedValues { get; set; }
    public List<string>? OneOf { get; set; }
    public List<string>? AnyOf { get; set; }
    public List<string>? AllOf { get; set; }
    public Dictionary<string, SchemaDefinition> Children { get; } = new();
    public List<string>? DependentRequired { get; set; }
    public List<string>? MutuallyExclusive { get; set; }
    public List<string>? ExactlyOne { get; set; }
    public SchemaCondition? Condition { get; set; }
    public List<(string key, List<object> values)>? Items { get; set; }

    public override string ToString() => $"SchemaDefinition({Type})";
}

/// <summary>
/// Represents a conditional schema structure (if/then/else).
/// </summary>
public record SchemaCondition(
    Dictionary<string, object> If,
    SchemaDefinition? Then,
    SchemaDefinition? Else);
